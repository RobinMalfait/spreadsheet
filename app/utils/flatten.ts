type Root<T> = T extends Array<infer U> ? Root<U> : T

export function* flatten<T>(values: T): Generator<Root<T>> {
  if (Array.isArray(values)) {
    for (let value of values) {
      yield* flatten(value)
    }
  } else {
    yield values as Root<T>
  }
}
