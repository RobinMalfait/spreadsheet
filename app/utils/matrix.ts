export function ensureMatrix<T>(value: T | T[] | T[][]): T[][] {
  if (!Array.isArray(value)) {
    return [[value]] as T[][]
  }

  if (!Array.isArray(value[0])) {
    return [value] as T[][]
  }

  return value as T[][]
}
