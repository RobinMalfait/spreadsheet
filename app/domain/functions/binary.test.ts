import { exampleTests } from '~/test/utils'
import * as functions from './binary'
await exampleTests(functions)
