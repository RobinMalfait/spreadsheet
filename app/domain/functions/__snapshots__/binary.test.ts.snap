// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@example tests > BIN2DEC(1101) 1`] = `
"
┌───┬────┐
│   │ A  │
├───┼────┤
│ 1 │ 13 │
└───┴────┘
"
`;

exports[`@example tests > BIN2HEX(1101) 1`] = `
"
┌───┬─────┐
│   │ A   │
├───┼─────┤
│ 1 │ "d" │
└───┴─────┘
"
`;

exports[`@example tests > BIN2OCT(111111111) 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ "777" │
└───┴───────┘
"
`;

exports[`@example tests > DEC2BIN(13) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ 1101 │
└───┴──────┘
"
`;

exports[`@example tests > HEX2BIN > HEX2BIN("1234") 1`] = `
"
┌───┬─────────────────┐
│   │ A               │
├───┼─────────────────┤
│ 1 │ "1001000110100" │
└───┴─────────────────┘
"
`;

exports[`@example tests > HEX2BIN > HEX2BIN("FF") 1`] = `
"
┌───┬────────────┐
│   │ A          │
├───┼────────────┤
│ 1 │ "11111111" │
└───┴────────────┘
"
`;

exports[`@example tests > OCT2BIN(777) 1`] = `
"
┌───┬─────────────┐
│   │ A           │
├───┼─────────────┤
│ 1 │ "111111111" │
└───┴─────────────┘
"
`;
